File: docs/source/templates/instance.py
"""MUSCLE3 Python component template.

Note that this template is not executable as is, please have a look at the
examples in ``docs/source/examples`` to see working components."""

from libmuscle import Instance, Message
from ymmsl import Operator

instance = Instance({
        Operator.F_INIT: ["F_INIT_Port"],
        Operator.O_I: ["O_I_Port"],
        Operator.S: ["S_Port"],
        Operator.O_F: ["O_F_Port"]})

while instance.reuse_instance():
    # F_INIT
    setting = instance.get_setting("setting")
    ...

    instance.receive("F_INIT_Port")
    ...

    while t_cur <= t_max:
        # O_I
        t_next = t + dt
        if t_next > t_max:
            t_next = None

        instance.send("O_I_Port", Message(t_cur, t_next, data))
        ...

        # S
        instance.receive("S_Port")
        ...

        t_cur += dt

    # O_F
    instance.send("O_F_Port", Message(t_cur, None, data))

File: docs/source/templates/instance.f90
program instance
    ! MUSCLE3 Fortran component template.
    !
    ! Note that this template is not executable as is, please have a look at the
    ! examples in ``docs/source/examples`` to see working components.

    use ymmsl
    use libmuscle
    implicit none

    type(LIBMUSCLE_PortsDescription) :: ports
    type(LIBMUSCLE_Instance) :: instance

    type(LIBMUSCLE_Message) :: rmsg, smsg
    type(LIBMUSCLE_Data) :: data

    real (selected_real_kind(15)) :: t_cur, t_max, dt


    ports = LIBMUSCLE_PortsDescription()
    call ports%add(YMMSL_Operator_F_INIT, 'F_INIT_Port')
    call ports%add(YMMSL_Operator_O_I, 'O_I_Port')
    call ports%add(YMMSL_Operator_S, 'S_Port')
    call ports%add(YMMSL_Operator_O_F, 'O_F_Port')
    instance = LIBMUSCLE_Instance(ports)
    call LIBMUSCLE_PortsDescription_free(ports)

    do while (instance%reuse_instance())
        ! F_INIT
        setting = instance%get_setting_as_real8('setting')
        ! ...

        rmsg = instance%receive('F_INIT')
        call LIBMUSCLE_Message_free(rmsg)
        ! ...

        do while (t_cur <= t_max)
            ! O_I
            smsg = LIBMUSCLE_Message(t_cur, data)
            if (t_cur + dt <= t_max) then
                call smsg%set_next_timestamp(t_cur + dt)
            end if
            call instance%send('O_I', smsg)
            call LIBMUSCLE_Message_free(smsg)
            ! ...

            ! S
            rmsg = instance%receive('S')
            call LIBMUSCLE_Message_free(rmsg)
            ! ...

            t_cur = t_cur + dt
        end do

        ! O_F
        smsg = LIBMUSCLE_Message(t_cur, data)
        call instance%send('final_state', smsg)
        call LIBMUSCLE_Message_free(smsg)
        ! ...

    end do

    call LIBMUSCLE_Instance_free(instance)

end program reaction

